### PR TITLE
Add expired payments

### PR DESCRIPTION
**Pull request by mhluska**
This PR adds an `expired` state to payments and cleans up the code a bit (change transition names to verbs etc).

@Sailias I wanted you to look this over before merge because I'm not 100% sure this is a common use case. My app has this use case because it shows a payment widget similar to Coinbase Commerce which will expire the payment after 15 minutes.

### CODE REVIEW

**Comment by Sailias** [comment]
@mhluska The commit looks ok but there are a couple of potential problems.  

I like that you:

* Made it so the payable can be notified of any specific event for any event through a generic method
* I like that you expire transactions, then check their payment status right after.  Just in case they have past the expiry time but have received a payment, you still process that.

Regarding the 2nd point, payments are going to expire a lot, and if we consider Bitcoin, the processing times will be long.  I feel that you should take an unconfirmed transaction into account when checking for your expiry time.

Perhaps an unconfirmed payment should be it's own state.  That way you don't need to worry about being too precise with an expiry time.

The goal would be to reduce a few things:

* number of times a paid payment is erroneously marked as expired
* number of database updates when processing transactions (update to expired, update to paid_in_full)
* need for the developer to fine tune the expiry time

Personally I don't see the need for a short expiry time.  The pricing processor provides a way to honour the fluctuating prices over time.  A longer expiry time, `3.days`, will dramatically decrease the number of API calls the system makes when checking for payments. (Right now it processes all payments that are pending payment since the beginning)  And lets face it, if they haven't submitted a transaction after 3 days, they probably aren't going to.

Either way, I think the flow should be the following:

```
if paid_in_full or partially_paid
elseif unconfirmed payment
elsif expired payment
end
```

Thoughts?

**Comment by mhluska** [comment]
@Sailias to your point about longer expiry times, I think you've found a mistake that I made. I intended this feature to be opt-in so no change in behavior unless `config.expire_payments_after` is defined. I'm fine with adding a longer delay to the readme/example code.

To your point about the flow you mentioned above, I think we're more or less on the same page. I set up the state machine so that only payments in the `pending` or `partially_paid` state can ever be expired:

```
transition [:pending, :partial_payment] => :expired
```

Any other `payment.expire` calls will just return false. In other words, as soon as there is a full payment with zero confirmations detected on the network, the payment can not expire anymore.

But we can update the code to make it impossible to expire partially paid payments as well:

```
transition [:pending] => :expired
```

Edit: Actually, now that I'm looking at this more closely, I see the bug you're pointing out. Maybe we can restructure the update logic like so:

```
def update_payment_state(payment)
  if payment.currency_amount_paid >= payment.price
    payment.pay
    if payment.transactions_confirmed?
      payment.confirm
    end
  elsif payment.currency_amount_paid > 0
    payment.partially_pay
  end

  expire_after = CryptocoinPayable.configuration.expire_payments_after
  if expire_after.present? && (Time.now - payment.created_at) >= expire_after
    payment.expire
  end
end
```

In other words, make the expire check at the end so that paid payments cannot enter the expired state.

**Comment by Sailias** [comment]
@mhluska of course, I'm only pointing out that shorter expiry_times create a window for the bug I mentioned above.

**Comment by mhluska** [comment]
@Sailias I'm not clear on the difference between the `paid_in_full` state and the proposed `unconfirmed` state.

**Comment by Sailias** [comment]
@mhluska right now when a transaction is processed there are only 3 possible states:

* paid_in_full
* partially_paid
* confirmed

I guess I'm advocating renaming that status, as a user might see it as confusing that it's "paid_in_full" but not "confirmed".

Perhaps this is just a UI issue where it could say:

* paid_in_full (1/3 confirmations)
* confirmed (6/3 confirmations)

In this scenario we wouldn't have to rename anything internally.

**Comment by Sailias** [comment]
@mhluska after looking at your modified code I see that you prevent a transaction from become expired using state machine as you only allow `pending => expired`.

Doing this check every time, then triggering `expire` when it can't happen because it's not in the correct state is a little wasteful and not really clear in the code, so it might be better to use an `if,elsif` to be more explicit.

```
def update_payment_state(payment)
  if payment.currency_amount_paid >= payment.price
    payment.pay
    if payment.transactions_confirmed?
      payment.confirm
    end
  elsif payment.currency_amount_paid > 0
    payment.partially_pay
  else
    expire_after = CryptocoinPayable.configuration.expire_payments_after
    if expire_after.present? && (Time.now - payment.created_at) >= expire_after
      payment.expire
    end
end
```

**Comment by mhluska** [comment]
Makes sense. I can update this.

**Comment by mhluska** [comment]
@Sailias this should be ready. Let me know what you think. I had to move the expired check after the payment processor runs because we don't want to move transactions into expired state before checking for updates from the source of truth.

File: lib/cryptocoin_payable/commands/payment_processor.rb [hunk 1]
```diff
@@ -4,6 +4,34 @@ def self.perform
       new.perform
     end
 
+    def self.update_transactions_for(payment)
```
Comment (mhluska) [inline lib/cryptocoin_payable/commands/payment_processor.rb]: I've moved this to its own public method because it may be useful to be able to explicitly sync transactions for a payment. In my own app I use BlockCypher's confidence interval web hook which allows me to immediately set a payment to confirmed state. But upon doing so, transactions should also be synced.

(Once this PR is merged I will add full support for BlockCypher web hooks to the gem)

File: lib/cryptocoin_payable/adapters/bitcoin.rb [hunk 2]
```diff
@@ -47,7 +46,7 @@ def self.create_address(id)
           blockHash: transaction['block_hash'],
           blockTime: transaction['confirmed'].nil? ? nil : DateTime.iso8601(transaction['confirmed']),
           estimatedTxTime: DateTime.iso8601(transaction['received']),
-          estimatedTxValue: transaction['outputs'].sum { |out| out['addresses'].join.eql?(address) ? out['value'] : 0 },
+          estimatedTxValue: transaction['outputs'].sum { |out| out['addresses'].present? && out['addresses'].join.eql?(address) ? out['value'] : 0 },
```
Comment (mhluska) [inline lib/cryptocoin_payable/adapters/bitcoin.rb]: In some rare cases `out['addresses']` was `nil` resulting in a runtime error.
